src/amo/withExperiment.js

```diff
@@ -155,19 +168,43 @@ export const withExperiment = ({
     constructor(props: withExperimentInternalProps) {
       super(props);
 
-      if (!this.isExperimentEnabled()) {
-        log.debug(`Experiment "${props.id}" is not enabled by config.`);
-        return;
+      const {
+        _config,
+        _getVariant,
+        _isExperimentEnabled,
+        cookies,
+        id,
+        variants,
+      } = this.props;
+
+      const registeredExperiments = this.getExperiments();
+
+      // Clear any disabled experiments from the cookie.
+      let cleanupNeeded = false;
+      const cleanedExperiments = { ...registeredExperiments };
+      for (const experimentId of Object.keys(registeredExperiments)) {
+        if (!_isExperimentEnabled({ _config, id: experimentId })) {
+          delete cleanedExperiments[experimentId];
+          cleanupNeeded = true;
+        }
       }
 
-      const { _getVariant, cookies, id, variants } = this.props;
+      if (cleanupNeeded) {
+        cookies.set(EXPERIMENT_COOKIE_NAME, cleanedExperiments, cookieConfig);
+      }
 
-      this.experimentCookie = cookies.get(this.getCookieName());
+      if (!_isExperimentEnabled({ _config, id })) {
+        log.debug(`Experiment "${props.id}" is not enabled by config.`);
+        return;
+      }
 
-      if (this.experimentCookie === undefined) {
+      if (!this.cookieIncludesExperiment(registeredExperiments)) {
         const variant = _getVariant({ variants });
-        this.experimentCookie = variant.id;
-        cookies.set(this.getCookieName(), this.experimentCookie, cookieConfig);
+        const newCookie = {
+          ...registeredExperiments,
+          [id]: variant.id,
+        };
+        cookies.set(EXPERIMENT_COOKIE_NAME, newCookie, cookieConfig);
```

Comment:
I may be missing something here, because I'm not familiar with cookie handing in React, but I read this as:
 - we clean out old experiments above, creating `cleanedExperiments`
 - then the cookie is written with `cleanedExperiments`
 - if we need to add a new experiment we create `newCookie`, but use `registeredExperiments` - _not_ `cleanedExperiments` - to create it
 - then write the cookie (potentially again)

so, questions - can we avoid writing the cookie twice if we need to both do clean up _and_ add a new experiment?  And shouldn't we be using `cleanedExperiments` to create `newCookie`?

Reply:
Good questions, and suggestions. The code evolved, which is why I am writing the cookie twice, but you're right, we should avoid that. As for the thing about `registeredExperiments` vs `cleanedExperiments` when adding the new experiment, what you are saying sounds correct, but I thought I had covered that with a test. I'll look into that as well.

I'm also going to make another change, moving the code that is currently in the `constructor` into the `render` method to deal with the fact that we cannot send GA events from the node server.

I'll flag you for another look once those things are done.